rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,438 @@
+the need for hard upgrades but this will be focused [elsewhere](https://github.com/tendermint/tendermint/issues/5595)).
+Instead, this RFC will describe what supporting soft upgrades would entail
+and explore two different methods: **Multi Data Structure Support** or
+**Chain Migrations**.
+
+## Proposal
+
+Tendermint supports a class of upgrades deemed as soft upgrades.
+
+The implementation to support this would ideally need to be done prior to the
+1.0 release.
+
+### Protocol Versioning
+
+Tendermint currently has three protocol versions: Block, P2P and App.
+
+The block protocol version would initially be the sole focus for soft upgrades.
+This consists of all block related data structures i.e. `Header`,
+`Vote`, `Validator`, `Commit` (a full list of the exact data structures can be
+found in [data structures](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md#data-structures).
+Data structures are dictated by the spec thus requiring a Tendermint software
+version to indicate the block version/s it supports. When a new spec release
+makes changes to any of the aforementioned data structures, this will result in
+incrementing the block protocol version.
+
+The app version can already be changed via consensus and is left to the
+discretion of the application to handle. The P2P version would also be left as
+it is, with major revisions requiring a hard upgrade. However, this may be
+revisited in the future.
+
+### Transitions
+
+Networks would be expected to announce upgrades in advance allowing a window for
+node operators to upgrade to a software version of Tendermint that would
+comply with the announced block protocol upgrade. This can be done
+asynchronously with node operators temporarily halting their node, changing to
+the new binary and starting up again. Tendermint could also offer a tool to
+easily swap between binaries thus reducing the amount of down-time that nodes
+experience although this is outside the scope of this RFC.
+
+The actual switch would occur in the same manner as an App Version change, using
+the `EndBlockResponse{}` ABCI call to indicate to Tendermint that a different
+block protocol be used. In the case that a node failed to upgrade, the node
+would gracefully stop and error.
+
+Given the delayed execution model, careful consideration would be needed to
+ensure that the nodes could transition smoothly. It may be that the actual
+upgrade might not take place till height: h + 2.
+
+As we expect upgrades to always increment the block version by one, rather
+than having the version number being passed in `EndBlockResponse{}`, another
+option could be to have the height where we want the upgrade to occur. With this
+approach validators are not reaching consensus on the version but the window of
+time (as a height) that nodes will have to perform or prepare for the upgrade.
+
+We will now cover the two main methods of executing a soft upgrade.
```

Comment:
It might help saying these two approaches on the opposite spectrum, and that we could also imagine having some kind of hybrid approaches that would be in between.